File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    const answer = [];
    // 문자열을 배열로 변환하여 방문 처리를 용이하게 함
    const grid = maps.map(row => row.split(''));
    const rows = grid.length;
    const cols = grid[0].length;
    
    // 상하좌우 이동을 위한 방향 벡터
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];

    // 지도 전체를 순회
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            // 바다가 아닌 땅(숫자)을 발견한 경우 BFS 시작
            if (grid[i][j] !== 'X') {
                let sum = 0; // 해당 섬의 총 식량
                let queue = [[i, j]];
                
                sum += Number(grid[i][j]);
                grid[i][j] = 'X'; // 방문한 곳은 'X'로 변경하여 중복 방문 방지

                while (queue.length > 0) {
                    const [cx, cy] = queue.shift();

                    // 상하좌우 4방향 탐색
                    for (let d = 0; d < 4; d++) {
                        const nx = cx + dx[d];
                        const ny = cy + dy[d];

                        // 지도 범위 내에 있고, 바다가 아닌 경우
                        if (nx >= 0 && nx < rows && ny >= 0 && ny < cols && grid[nx][ny] !== 'X') {
                            sum += Number(grid[nx][ny]);
                            grid[nx][ny] = 'X'; // 방문 처리
                            queue.push([nx, ny]);
                        }
                    }
                }
                // 탐색이 종료되면 합산된 식량을 배열에 추가
                answer.push(sum);
            }
        }
    }

    // 섬이 존재하면 오름차순 정렬, 존재하지 않으면 [-1] 반환
    return answer.length > 0 ? answer.sort((a, b) => a - b) : [-1];
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque

def solution(maps):
    rows = len(maps)
    cols = len(maps[0])
    visited = [[False] * cols for _ in range(rows)]
    answer = []
    
    # 상, 하, 좌, 우 이동 방향
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    for i in range(rows):
        for j in range(cols):
            # 바다가 아니고 아직 방문하지 않은 땅인 경우 BFS 시작
            if maps[i][j] != 'X' and not visited[i][j]:
                q = deque([(i, j)])
                visited[i][j] = True
                total_food = int(maps[i][j])
                
                # BFS 탐색
                while q:
                    r, c = q.popleft()
                    
                    for dr, dc in directions:
                        nr = r + dr
                        nc = c + dc
                        
                        # 지도 범위를 벗어나지 않는지 확인
                        if 0 <= nr < rows and 0 <= nc < cols:
                            # 연결된 땅이면서 방문하지 않았다면 큐에 추가
                            if maps[nr][nc] != 'X' and not visited[nr][nc]:
                                visited[nr][nc] = True
                                total_food += int(maps[nr][nc])
                                q.append((nr, nc))
                                
                answer.append(total_food)
                
    # 무인도가 없는 경우 -1 반환
    if not answer:
        return [-1]
        
    # 오름차순 정렬하여 반환
    return sorted(answer)